### PR TITLE
Add BridgedEngine to tabs

### PR DESCRIPTION
**Pull request by mhammond**
This is for tabs on desktop. I can't think of anything outstanding left to do! No hurry though - there's no need to land this until the matching desktop patch is ready.

### CODE REVIEW

File: components/tabs/src/sync/bridge.rs
Comment (mhammond) [reply components/tabs/src/sync/bridge.rs:184]: We could do the types too, but in many ways I'd rather think about exposing them directly from sync15

However, it doesn't exist *just* to convert the types - uniffi expects a struct here, but BridgedEngine is a trait - so avoid this entirely would involve *both* the types described in UniFFI, but also the ability to tell UniFFI that BridgedEngine is actually a trait (which is what we've been talking about on and off and I hope to get back to soon, but probably not in time to have it before we want to merge this).

Please correct me if I'm wrong though, because I quite hate this :)

File: components/tabs/src/sync/sync_impl.rs
Comment (mhammond) [reply components/tabs/src/sync/sync_impl.rs:82]: Hopefully the recent sync15 work makes this more obvious. Once #5139 we can be far more aggressive here, and I can probably move this "impl" back into SyncEngine, and BridgedEngine can just have a SyncEngine and build an IncomingChangeset etc - I started to play with that in my current version, but the Payload vs bridged_engine::IncomingEnvelope added just enough friction that I abandoned it.

**Comment by mhammond** [comment]
I think this new version is quite a bit saner now that I've merged sync15-traits and sync15. What it means is that *both* Desktop and Mobile get *both* SyncEngine and BridgedEngine, which simplifies things quite a bit, and the overhead of having both even though each platform will only use 1 seems low enough that the simplicity is worthwhile.

If #5139 lands before this needs to it would become *far* cleaner (we could get rid of SyncImpl) but that seems unlikely at this stage - but even if approved I'll probably hold off merging until I'm sure that's true.

File: components/tabs/src/sync/bridge.rs
Comment (bendk) [reply components/tabs/src/sync/bridge.rs:185]: I think that's right, but let me double check my understanding.  The the path forward would be:

  - Expose `IncomingEnvelope` in `sync15.udl` and implement external types on desktop.  We could then rename `BridgedEngineImpl` to `TabsBridgedEngine` and expose that directly without the need for this wrapper struct.
  - Add support for trait objects in UniFFI, expose the `BridgedEngine` trait in `sync15.udl`, then `tabs.udl` doesn't need to spell out the methods, it can just say that `TabsStore.bridged_engine()` returns a `BridgedEngine`.

(Not necessarily saying that this would need to be done in 2 PRs, maybe it would be simpler to do them both in 1).

**Review by bendk — APPROVED** [review]
Still looking good to me.

**Comment by mhammond** [comment]
>    Add support for trait objects in UniFFI, expose the BridgedEngine trait in sync15.udl, then tabs.udl doesn't need to spell out the methods, it can just say that TabsStore.bridged_engine() returns a BridgedEngine.

Yep, exactly. I'm not even sure this would require external types for desktop - all new types would be local to sync15.udl, and with uniffi growing support for traits, I don't think the fact the trait is implemented in a different crate would force the existing external types mechanisms to be used?

(OTOH though, there are duplicated Device types involved, so I think we'd be made to do this without also cleaning that up, which would need external types)

**Review by skhamis — APPROVED** [review]
Awesome job with this -- LGTM! 👍

**Pull request merged**
